rust/src/client/mobile_client/client.rs

```diff
@@ -0,0 +1,364 @@
+        info!("new participant with undefined task");
+        if let Err(err) = self.fetch_round_params().await {
+            error!("{:?}", err);
+            return self.reset().into();
+        };
+
+        let Self {
+            proxy,
+            round_params,
+            participant,
+            local_model,
+            global_model,
+        } = self;
+
+        let participant_type = participant.determine_type(
+            round_params.seed.as_slice(),
+            round_params.sum,
+            round_params.update,
+        );
+
+        match participant_type {
+            Type::Unselected(unsel_par) => {
+                info!("unselected");
+                ClientState::<Undefined>::new(proxy, unsel_par.reset(), local_model, global_model)
+                    .into()
+            }
+            Type::Summer(sum_par) => {
+                ClientState::<Sum>::new(proxy, round_params, sum_par, local_model, global_model)
+                    .into()
+            }
+            Type::Updater(upt_pat) => {
+                ClientState::<Update>::new(proxy, round_params, upt_pat, local_model, global_model)
+                    .into()
+            }
+        }
+    }
+
+    async fn fetch_round_params(&mut self) -> Result<(), ClientError> {
+        self.round_params = self.proxy.get_round_params().await?;
+        self.fetch_global_model().await;
+        Ok(())
+    }
+
+    async fn fetch_global_model(&mut self) {
+        if let Ok(model) = self.proxy.get_model().await {
+            //update our global model where necessary
+            let mut global_model = self.global_model.borrow_mut();
+
+            match (model, global_model.as_ref()) {
+                (Some(new_model), None) => {
+                    info!("new global model");
+                    *global_model = Some(new_model);
+                }
+                (Some(new_model), Some(old_model)) if &new_model != old_model => {
+                    debug!("updating global model");
+                    *global_model = Some(new_model);
+                }
+                (None, _) => debug!("global model not ready yet"),
+                _ => debug!("global model still fresh"),
+            }
+        }
+    }
+}
+
+impl ClientState<Sum> {
+    fn new(
+        proxy: Proxy,
+        round_params: RoundParameters,
+        participant: Participant<Sum>,
+        local_model: Rc<RefCell<Option<Model>>>,
+        global_model: Rc<RefCell<Option<Model>>>,
+    ) -> Self {
+        Self {
+            proxy,
+            round_params,
+            participant,
+            local_model,
+            global_model,
+        }
+    }
+
+    async fn next(mut self) -> ClientStateMachine {
+        info!("selected to sum");
+
+        match self.run().await {
+            Ok(_) => self.move_into_sum2().into(),
+            Err(ClientError::RoundOutdated) => self.reset().into(),
+            Err(err) => {
+                error!("{:?}", err);
+                self.into()
```

Comment:
I think we need a timer for the error case. Because if I understand the implementation correctly, if some resource is not available yet, the client we'll keep querying it and fail with `TooEarly` in a sort of busy loop. On the longer run we might even want to have a more sophisticated retry mechanism with exponential backoff or something.

Reply:
I agree that is a good idea.
Another idea I had is to add caching to make the client more efficient. At the moment we always request all values even if we already got a value in the previous call of `next()`.